refactor, docs: Embedded ASMap [2/3]: Refactor asmap internals and add documentation

fjahr opened the pull request:
This is a second slice carved out of #28792. It contains the following changes that are crucial for the embedding of asmap data which is added the following PR in the series (probably this will remain in #28792).

The changes are:
- Modernizes and simplifies the asmap code by operating on `std::byte` instead of bits
- Unifies asmap version calculation and naming (previously it was called version and checksum interchangeably)
- Operate on a `span` rather than a vector in the asmap internal to prevent holding the asmap data in memory twice
- Add more extensive documentation to the asmap implementation
- Unify asmap casing in implemetation function names

The first three commits were already part of #28792, the others are new.

The documentation commit came out of feedback gathered at the latest CoreDev. The primary input for the documentation was the documentation that already existed in the Python implementation (`contrib/asmap/asmap.py`) but there are several other comments as well. Please note: I have also asked several LLMs to provide suggestions on how to explain pieces of the implementation and better demonstrate how the parts work together. I have copied bits and pieces that I liked but everything has been edited further by me and obviously all mistakes here are my own.

fjahr committed:
refactor: Operate on bytes instead of bits in Asmap code

Co-authored-by: Hodlinator <172445034+hodlinator@users.noreply.github.com>

fjahr committed:
refactor: Unify asmap version calculation and naming

Calculate the asmap version only in one place: A dedicated function in util/asmap.

The version was also referred to as asmap checksum in several places. To avoid confusion call it asmap version everywhere.

fjahr committed:
refactor: Use span instead of vector for data in util/asmap

This prevents holding the asmap data in memory twice.

The version hash changes due to spans being serialized without their size-prefix (unlike vectors).

fjahr committed:
doc: Add more extensive docs to asmap implementation

Also makes minor improvement on the python implementation documentation.

sipa commented on src/util/asmap.cpp:
```diff
@@ -21,140 +21,235 @@
 #include <vector>
 
+/*
+ * ASMap (Autonomous System Map) Implementation
```
All the documentation changes in this commit look correct to me.

One thing that may be worth mentioning early on is that it's a bit-packed format, i.e., the entire compressed mapping is treated as a sequence of bits, packed together at 8 per byte, which are concatenated encodings of instructions.

fjahr committed:
refactor: Simplify Interpret asmap function

This aligns it more with SanityCheckAsmap and reduces variable scope.

Also unify asmap casing in SanityCheckAsmap function name.

Co-authored-by: Hodlinator <172445034+hodlinator@users.noreply.github.com>

hodlinator commented on src/netgroup.cpp:
```diff
@@ -81,31 +83,28 @@ std::vector<unsigned char> NetGroupManager::GetGroup(const CNetAddr& address) co
-                ip_bits[byte_i * 8 + bit_i] = (IPV4_IN_IPV6_PREFIX[byte_i] >> (7 - bit_i)) & 1;
-            }
+            ip_bits[byte_i] = std::byte(IPV4_IN_IPV6_PREFIX[byte_i]);
         }
```
The `for` doing `ip_bits[byte_i] = std::byte(IPV4_IN_IPV6_PREFIX[byte_i]);` can now be removed since we have `copy_n` doing the same thing above.

hodlinator commented on src/util/asmap.cpp:
```diff
@@ -203,6 +305,10 @@ bool SanityCheckASMap(const std::span<const std::byte>& asmap, int bits)
 
+/**
+ * Provides a safe interface for validating ASMap data before use. Validates
+ * the data and returns it if valid, empty span otherwise.
```
In 0196f70a8c99b9c03c81e879cd9f05560fc20543: Could replace second sentence with "Returns true if the data is valid for 128 bits long inputs" or just remove it (no longer returns span).

hodlinator commented on src/util/asmap.h:
```diff
@@ -5,16 +5,23 @@
 #include <vector>
 
-uint32_t Interpret(const std::vector<bool> &asmap, const std::vector<bool> &ip);
+uint32_t Interpret(std::span<const std::byte> asmap, std::span<const std::byte> ip);
```
Continuing https://github.com/bitcoin/bitcoin/pull/28792#discussion_r2514801922

We still mixed `std::span` and `const std::span&` here in:

https://github.com/bitcoin/bitcoin/blob/eaf5e12d2b52d66a75a68cf94d328d0ec9048986/src/util/asmap.h#L16-L25

hodlinator reviewed (commented):
Reviewed eaf5e12d2b52d66a75a68cf94d328d0ec9048986

Thanks for peeling off this PR as I suggested!

fjahr commented on src/util/asmap.cpp:
```diff
@@ -5,186 +5,297 @@
+    // It is followed by an integer using jump encoding.
     JUMP = 1,
+    // A match instruction, encoded as [1,1,0], inspects 1 or more of the next unused bits
+    // in the input. If they all match, execution continues. If not, the default ASN is returned.
```
Done. And I just clarified that failure means 0 in python, otherwise there would have been more edits necessary to make this clear, for example the line below your suggestion still talks about a failure too.

hodlinator commented on src/test/netbase_tests.cpp:
```diff
@@ -625,19 +625,10 @@ BOOST_AUTO_TEST_CASE(asmap_test_vectors)
 
     // Construct NetGroupManager with this data.
-    NetGroupManager netgroup{std::move(asmap_bits)};
+    NetGroupManager netgroup{std::move(ASMAP_DATA)};
```
I meant you could could have this in the intermediate commit:

```C++
NetGroupManager netgroup{std::vector(ASMAP_DATA.begin(), ASMAP_DATA.end())};
```
But then once we reach the commit adding embedding support change to:
```C++
auto netgroup{NetGroupManager::WithEmbeddedAsmap(ASMAP_DATA)};
```
...ending up going directly from `std::array` -> `std::span`.

---

If you prefer the current way I would still avoid `auto` and change:

```diff
- auto asmap_vec{std::vector(ASMAP_DATA.begin(), ASMAP_DATA.end())};
+ std::vector asmap_vec(ASMAP_DATA.begin(), ASMAP_DATA.end());
  NetGroupManager netgroup{asmap_vec};
```
But no big deal.

hodlinator replied:
Thank you! In adfd800626a7910a603c34ab733fc1e91d438bc5 the `ASMAP_DATA` type is still unnecessarily changed from `constexpr auto` to `static const auto`.

hodlinator commented on src/util/asmap.cpp:
```diff
@@ -5,186 +5,297 @@
+            uint32_t jump_offset = pos + jump;
             if (!jumps.empty() && jump_offset >= jumps.back().first) return false; // Intersecting jumps
-            jumps.emplace_back(jump_offset, bits);
+            jumps.emplace_back(jump_offset, bits);  // Queue jump target for validation
```
TODO: understand jumps

hodlinator commented on src/util/asmap.cpp:
```diff
@@ -5,186 +5,297 @@
+ */
+inline bool GetBitBE(std::span<const std::byte> bytes, uint32_t bitpos) noexcept
+{
+    return (std::to_integer<uint8_t>(bytes[(bytes.size() * 8 - bitpos) / 8]) >> (7 - ((bytes.size() * 8 - bitpos) % 8))) & 1;
```
There's an reduced version that seems to do the same
```suggestion
    return (std::to_integer<uint8_t>(bytes[(bytes.size() * 8 - bitpos) / 8]) >> ((7 + bitpos) % 8)) & 1;
```

---

However, I'm surprised by the way the shift amount changes for different bit positions. Would have expected something closer to...

```suggestion
    return (std::to_integer<uint8_t>(bytes[(bytes.size() * 8 - bitpos) / 8]) >> (7 - (bitpos % 8))) & 1;
```
....but it fails the unit tests.

Godbolt experiment https://godbolt.org/z/4Tc34hcdG outputs:
```
bitpos: 0, PR bit shift: 7, reduced PR: 7, expected BE bit shift: 7
bitpos: 1, PR bit shift: 0, reduced PR: 0, expected BE bit shift: 6
bitpos: 2, PR bit shift: 1, reduced PR: 1, expected BE bit shift: 5
bitpos: 3, PR bit shift: 2, reduced PR: 2, expected BE bit shift: 4
bitpos: 4, PR bit shift: 3, reduced PR: 3, expected BE bit shift: 3
bitpos: 5, PR bit shift: 4, reduced PR: 4, expected BE bit shift: 2
bitpos: 6, PR bit shift: 5, reduced PR: 5, expected BE bit shift: 1
bitpos: 7, PR bit shift: 6, reduced PR: 6, expected BE bit shift: 0
bitpos: 8, PR bit shift: 7, reduced PR: 7, expected BE bit shift: 7
bitpos: 9, PR bit shift: 0, reduced PR: 0, expected BE bit shift: 6
bitpos: 10, PR bit shift: 1, reduced PR: 1, expected BE bit shift: 5
bitpos: 11, PR bit shift: 2, reduced PR: 2, expected BE bit shift: 4
bitpos: 12, PR bit shift: 3, reduced PR: 3, expected BE bit shift: 3
bitpos: 13, PR bit shift: 4, reduced PR: 4, expected BE bit shift: 2
bitpos: 14, PR bit shift: 5, reduced PR: 5, expected BE bit shift: 1
bitpos: 15, PR bit shift: 6, reduced PR: 6, expected BE bit shift: 0
```

hodlinator commented on src/util/asmap.cpp:
```diff
@@ -5,186 +5,297 @@
+    const size_t endpos{asmap.size() * 8};
+    uint8_t bits = ip.size() * 8;
     uint32_t default_asn = 0;
     uint32_t jump, match, matchlen;
```
nit: Here are some changes that bring the `Interpret()` implementation even closer to `SanityCheckAsmap()` while also decreasing variable scope, making the code more readable. Maybe something to include in the last commit which aligns the casing of `SanityCheckAsmap`?
```diff
--- a/src/util/asmap.cpp
+++ b/src/util/asmap.cpp
@@ -183,18 +183,16 @@ uint32_t Interpret(const std::span<const std::byte> asmap, const std::span<const
     const size_t endpos{asmap.size() * 8};
     uint8_t bits = ip.size() * 8;
     uint32_t default_asn = 0;
-    uint32_t jump, match, matchlen;
-    Instruction opcode;
     while (pos < endpos) {
-        opcode = DecodeType(pos, asmap);
+        Instruction opcode = DecodeType(pos, asmap);
         if (opcode == Instruction::RETURN) {
             // Found leaf node - return the ASN
-            default_asn = DecodeASN(pos, asmap);
-            if (default_asn == INVALID) break; // ASN straddles EOF
-            return default_asn;
+            uint32_t asn = DecodeASN(pos, asmap);
+            if (asn == INVALID) break; // ASN straddles EOF
+            return asn;
         } else if (opcode == Instruction::JUMP) {
             // Binary branch: if IP bit is 1, jump forward; else continue
-            jump = DecodeJump(pos, asmap);
+            uint32_t jump = DecodeJump(pos, asmap);
             if (jump == INVALID) break; // Jump offset straddles EOF
             if (bits == 0) break; // No input bits left
             if (int64_t{jump} >= static_cast<int64_t>(endpos - pos)) break; // Jumping past EOF
@@ -208,11 +206,11 @@ uint32_t Interpret(const std::span<const std::byte> asmap, const std::span<const
             // The match value encodes both length and pattern:
             // - highest set bit position determines length (bit_width - 1)
             // - lower bits contain the pattern to compare
-            match = DecodeMatch(pos, asmap);
+            uint32_t match = DecodeMatch(pos, asmap);
             if (match == INVALID) break; // Match bits straddle EOF
-            matchlen = std::bit_width(match) - 1;  // An n-bit value matches n-1 input bits
+            int matchlen = std::bit_width(match) - 1;  // An n-bit value matches n-1 input bits
             if (bits < matchlen) break; // Not enough input bits
-            for (uint32_t bit = 0; bit < matchlen; bit++) {
+            for (int bit = 0; bit < matchlen; bit++) {
                 if (GetBitBE(ip, bits) != ((match >> (matchlen - 1 - bit)) & 1)) {
                     return default_asn;  // Pattern mismatch - use default
                 }
```

hodlinator reviewed (commented):
Reviewed b41f5a29f7d56da8fd157770ad29b5776c3684c6

Sorry for not finding these earlier, getting closer to A-C-K.

fjahr commented on src/util/asmap.h:
```diff
@@ -17,5 +18,7 @@ bool SanityCheckASMap(const std::vector<std::byte>& asmap, int bits);
 /** Read asmap from provided binary file */
 std::vector<std::byte> DecodeAsmap(fs::path path);
+/** Calculate the asmap version, a checksum identifying the asmap being used. */
+uint256 AsmapVersion(const std::vector<std::byte>& data);
```
I actually slightly prefer to keep it as is, because if I do this earlier I also have to change the tests because the hash result changes too with the span usage. I think it makes more sense to have this in the later commit where there is a much bigger overall benefit of introducing span evident that is actually what justifies the hash result being changed.

hodlinator replied:
Ah, makes sense. :+1:

fjahr commented on src/util/asmap.h:
```diff
@@ -10,15 +10,18 @@
+/** Read and check asmap from provided binary file */
 std::vector<std::byte> DecodeAsmap(fs::path path);
+/** Check asmap from embedded data */
+bool CheckAsmap(std::span<const std::byte> data);
```
Renamed to `CheckStandardAsmap` and using it in a two more places, but keeping it out of asmap_direct fuzzer because we are testing the bits there explicitly. Moved it below below the sanity check in .h which is the same location as .cpp

fjahr commented on src/util/asmap.cpp:
```diff
@@ -5,186 +5,297 @@
+ * Extract a single bit from byte array using big-endian bit ordering.
+ * Used for IP addresses to match network byte order conventions.
+ */
+inline bool GetBitBE(std::span<const std::byte> bytes, uint32_t bitpos) noexcept
```
Thanks, I have applied this latest patch, I think that also resolves your earlier comment that was in a different comment thread.

fjahr commented:
Addressed the latest review comments from @hodlinator , thanks for the detailed patch suggestions and rationale, co-authorship added!

hodlinator reviewed (approved):
ACK 825d677a863d153d4c466288397da773ee2bb2f5

Operating on `std::span<std::byte>` rather than `std::vector<bool>&` enables for efficient use of asmap data embedded in the binary (#28792).

I have thoroughly reviewed and contributed patches for this refactoring up to the point of having a conflict of interest in getting it merged. :)

To build more confidence of the `bool` -> `byte` transition, I developed a C++ test with more readable asmap input data - cf34bb9806e95116d4b65b6dc80dc11a1f723a51 (https://github.com/hodlinator/bitcoin/tree/pr/33878_suggestions2 for interleaved branch). Not sure whether it's worth including in some form.

fjahr commented:
> To build more confidence of the bool -> byte transition, I developed a C++ test with more readable asmap input data - https://github.com/bitcoin/bitcoin/commit/cf34bb9806e95116d4b65b6dc80dc11a1f723a51 (https://github.com/hodlinator/bitcoin/tree/pr/33878_suggestions2 for interleaved branch). Not sure whether it's worth including in some form.

Thanks, looks like a very nice visualization at least. I will consider including it and addressing your other comments until I have to retouch here. If that doesn't happen there are follow-ups already planned where these could be included. Thanks again for the reviews :)

sedited commented on src/netgroup.h:
```diff
@@ -70,8 +86,19 @@ class NetGroupManager {
+     * asmap data.
+     */
+    const std::span<const std::byte> m_asmap;
+    std::vector<std::byte> m_loaded_asmap;
```
In commit dc0c4310bae257236d3cd6a06de8071276d4548b

You mention that this dual ownership structure is done to avoid holding the asmap in memory twice. Where is this currently the case? Does that only happen in test code at the moment?

fjahr replied:
It was the case an earlier version of the code. I guess referencing something that isn't actually there anymore isn't something we should be doing in docs, even if it's implicitly. But in this case I think there should be some documentation in this regard so it clears up why we need `m_asmap` and `m_loaded_asmap` here.

fjahr commented on src/util/asmap.cpp:
```diff
@@ -214,19 +222,18 @@ std::vector<std::byte> DecodeAsmap(fs::path path)
     }
     int64_t length{file.size()};
     LogInfo("Opened asmap file %s (%d bytes) from disk", fs::quoted(fs::PathToString(path)), length);
-
```
done

fjahr commented:
Addressed comments by @hodlinator and @sedited

hodlinator reviewed (approved):
re-ACK c5251f39a90ff63bed417b0113ba4fdbc7ea78d0

Changes since my first ACK (https://github.com/bitcoin/bitcoin/pull/33878#pullrequestreview-3503770177) are only minor adjustments based off that review and another (https://github.com/bitcoin/bitcoin/pull/33878#pullrequestreview-3615845587):
* Made `NetGroupManager`-ctor temporarily take r-value
* Avoided span-references
* Cleaned up whitespace across commits

sedited reviewed (approved):
ACK c5251f39a90ff63bed417b0113ba4fdbc7ea78d0

hodlinator reviewed (approved):
re-ACK 4fec726c4d352daf2fb4a7e5ed463e44c8815ddb

Changes since previous ACK (https://github.com/bitcoin/bitcoin/pull/33878#pullrequestreview-3624323262):
* Resolved conflict with fabf8d1c5bdb6a944762792cdc762caa6c1a760b

sipa reviewed (commented):
ACK 4fec726c4d352daf2fb4a7e5ed463e44c8815ddb

sedited reviewed (approved):
Re-ACK 4fec726c4d352daf2fb4a7e5ed463e44c8815ddb
